### DISCUSSION TITLE
Plugins Management

### DISCUSSION THREAD

**Discussion by lfbdev** [devstream-io/devstream#26] [Show and tell]
1. .so files can be uploaded to github repo assets, and can be downloaded via github api
![image](https://user-images.githubusercontent.com/94163234/144823995-5fbce698-e9c0-41b9-afd2-d6867275aeba.png)

2. Also, we can set the retries param by using go-resty project, and can be callback while condition triggered
![image](https://user-images.githubusercontent.com/94163234/144824241-36aaba9c-9a56-44c4-aee8-efd2509c8b53.png)

3.Some Plugins are very large, such as argocd.so, even over 60M. When downloaded from assets, it may cost 5 mins or above.(both via website or api.gihub.com), so the user feeling is poor

**Comment by IronCore864** [comment]
I have also done some research, and here's the result:

## 1. `go-resty`

An excellent choice. Fast, easy to use, detailed documentation, lots of stars.

## 2. Downlaod

We do not need to use GitHub API for downloading a release. This would make things more complicated, like we have to figure out the asset ID, and set the token.

We can use the download URL which we can copy from the release page, for example:

```go
	resp, err := client.R().SetOutput("v3.68.0.tar.gz").
		Get("https://github.com/hashicorp/terraform-provider-aws/archive/refs/tags/v3.68.0.tar.gz")
```

With a not-so-good proxy, I can download about 50MB in about 30 seconds, which I think is more or less okayish:

```bash
$ ./test 
2021-12-07 11:50:26.3664257 +0800 CST m=+0.000358201
200
2021-12-07 11:50:51.6884409 +0800 CST m=+25.322373401
$ ls -lh v3.68.0.tar.gz 
-rw-r--r-- 1 tiexin tiexin 49M Dec  7 11:50 v3.68.0.tar.gz
```

## 3 Retry/Timeout

I think it's relatively difficult to set a reasonable timeout because different plugins have different sizes. Let's forget about a timeout for now; retry is still necessary.

Maybe it's possible to read the size before download and calculate a reasonable timeout value; this could be a future improvement. Showing a progress bar is even better.

Please proceed with a PR @lfbdev; I will put more information in the corresponding ticket in the project.

@daniel-hutao please also have a look.

**Comment by IronCore864** [comment]
Ticket updated: https://github.com/merico-dev/stream/issues/17

Very well done on the investigation. Kudos!
